### ISSUE TITLE
Expose source mapping through JSGenApi

### ISSUE THREAD

**Issue by benmerckx** [HaxeFoundation/haxe#8625]
Would love to get access to source mapping data for generated code by JSGenApi (generateValue & generateStatement).

**Comment by Aurel300** [comment]
Both `generateValue` and `generateStatement` take a `TypedExpr`, which has a `pos` field, expressing where in the source code this expression comes from. Is this not what you want?

**Comment by benmerckx** [comment]
Yes, but I'm looking to get that information for every sub expression as well. I thought of going through the expressions but it essentially means reimplementing gen_exp and gen_value of genjs.

**Comment by Aurel300** [comment]
[`TypedExprTools`](https://api.haxe.org/v/development/haxe/macro/TypedExprTools.html) has `iter` (and `map`), which lets you provide a function that is called on every sub-expression of a given `TypedExpr`.

**Comment by benmerckx** [comment]
That would help iterating and finding source positions. But it doesn't help that much finding generated code positions (unless I'm not thinking this through). In the scenario where I'd use JSGenApi to generate expressions I can get the code, then iterate through sub expressions. But to have an idea of position in the generated code (which was already generated for it's parent expr) I'd have to generate code again and then somehow string search its position in the parent code.

**Comment by Aurel300** [comment]
Oh right, now I understand. I'm not sure this can be made easily accessible. The actual positions in the generated code can only be determined once the code is generated, since a change in any expression results in everything after it in the file changing positions as well …

Perhaps you could insert `/*POS:...*/` markers in the generated code, then post-process that? I'm not sure what your intended application is.

**Comment by benmerckx** [comment]
> I'm not sure what your intended application is.

I was looking into generating split js files(#5831). [Hxgenjs](https://github.com/kevinresol/hxgenjs) does just that using a custom generator but provides no sourcemap support. I realize adding it involves tracking all generated code positions. Couldn't see where to get at that data for `generateValue` though. Inserting the position data into the code would work but I fear that would slow the whole process down a lot more.
